woocommerce/woocommerce-ios
# Add support for WP-API error parsing

Closes: #8030 


### Description


The existing networking code uses `DotcomValidator` and `DotcomError` to attempt to extract error information from WordPress.com API responses (`error`/`message`). However, the newer endpoints that use the core REST API (`wpcom/v2`, `wp/v2`) use a different error format (`code`/`message`/`data`).

This PR got a bit complex, but the main idea is that each type of request can define a validator that might convert a response into an error. This is useful because a `DotcomRequest` might want to do a different thing depending on the `wordpressApiVersion`.

### Testing instructions


The easiest way to force an error is to modify `InAppPurchasesRemote` and make sure the `X-APP-ID` header has an unknown identifier (e.g. `Bundle.main.bundleIdentifier?.appending(".unknown")`).

Then:

1. Go to Hub menu > Debug IAP
2. Wait for the products to load
3. Look at the console for an error

If you test this on trunk, all you would see is a parsing error:

> Error loading products: typeMismatch(Swift.Array<Any>, Swift.DecodingError.Context(codingPath: [], debugDescription: "Expected to decode Array<Any> but found a dictionary instead.", underlyingError: nil))

With the changes in this PR, you should see a more specific error:

> Error loading products: WordPress API Error: [rest_forbidden] Sorry, you are not allowed to do that.

---
